### DISCUSSION TITLE
MU-MIMO simulations with differing CP and Tx antenna configurations

### DISCUSSION THREAD

**Discussion by naperman** [NVlabs/sionna#428] [General]
I'm trying to simulate the following MU-MIMO scenarios:

1) Some UT's and interferers on the cell edge(uses extended CP) and some closer to the BS(uses normal CP). 
2) UTs with differing number of layers, codebooks and antennas

The first scenario is more important than the second. 

The sionna.nr.PUSCHTransmitter and sionna.nr.PUSCHReceiver API accepts transmit and received waveforms of the following dimensions:

PUSCHTransmitter
```
Time-domain input signal:
x ([batch size, num_tx, num_tx_ant, num_ofdm_symbols, fft_size], tf.complex 
or
or
Frequency-domain input signal:
 [batch size, num_tx, num_tx_ant, num_time_samples], tf.complex)
```

PUSCHReceiver
```
Time-domain input signal:
y ([batch size, num_rx, num_rx_ant, num_ofdm_symbols, fft_size], tf.complex 
or
Frequency-domain input signal:
[batch size, num_rx, num_rx_ant, num_time_samples + l_max - l_min], tf.complex) – Frequency- or time-domain input signal
```

So from the above, I understand 'num_ofdm_symbols' and 'num_tx_ant' have to be consistent across the UTs and interferers. Is there any workaround to support differing UT and interferer configurations?

**Comment by SebastianCa** [comment]
Hi @naperman,

This is correct, the output dimensions must be consistent across UTs. A simple workaround is to instantiate multiple `PUSCHTransmitter` instances. Since the interferer configuration is typically unknown to the `PUSCHReceiver`, it should suffice to associate only the `PUSCHTransmitter` of the UT with the `PUSCHReceiver`. Note that you might need to pad the transmit signals before concatenating or superimposing the tensors.

Alternatively, masking might also be an option.

I hope this helps.

**Comment by APoornima** [comment]
Hi, I want to simulate MU-MIMO, I have SISO SCFDE and DFTsOFDM simulator, what is the best way to extend my SISO simulator to downlink MU-MIMO scenario efficiently?
Thank you in advance

**Reply by SebastianCa** [reply]
Hi @APoornima,
I would recommend to have a look at the [MIMO OFDM Tutorial](https://nvlabs.github.io/sionna/examples/MIMO_OFDM_Transmissions_over_CDL.html) and the [Multi-User MIMO Tutorial](https://nvlabs.github.io/sionna/examples/Realistic_Multiuser_MIMO_Simulations.html). These notebooks should help to get started with MIMO simulations.

**Reply by APoornima** [reply]
thank you.
I'm not sure if I have the stream correctly, I am simulating downlink scenario, Number of Tx antennas at base station = 32, number of UTs = 4, each UE has one antenna.

 NUM_TX = 32
NUM_RX = 4
batch_size = 1

NUM_STREAMS_PER_TX = 32 #BS with 4 antenna each to each user
RX_TX_ASSOCIATION = np.zeros([NUM_RX, NUM_TX]) #receiver, transmitter
RX_TX_ASSOCIATION[:, 0] = 1 #One BS send one stream to each receiver

STREAM_MANAGEMENT = sionna.mimo.StreamManagement(RX_TX_ASSOCIATION, NUM_STREAMS_PER_TX)

**Comment by SebastianCa** [comment]
I think that is not exactly what you want.
- As you have only one BS, it should be `num_tx=1` (with `num_tx_ant=32`). 
- The stream_managment should associate the transmitter to each receiver, i.e., `np.ones(num_tx=1, num_rx=4)`
- I assume you want one stream to each receiver, i.e., `NUM_STREAMS_PER_TX=1`

**Reply by APoornima** [reply]
thank you for the clarification, I am very new to this concept, hence finding it hard :)
